File: 1. Creational Design Patterns/Singleton/C#/SingletonExample.cs
// Singleton Pattern Example in C#
// Thread-safe, lazy initialization using .NET's Lazy<T>.
// Compile: csc SingletonExample.cs && SingletonExample.exe

using System;

sealed class AppConfig
{
    private static readonly Lazy<AppConfig> _instance = new(() => new AppConfig());
    public static AppConfig Instance => _instance.Value;

    // Example state
    public string ConnectionString { get; private set; }

    private AppConfig()
    {
        ConnectionString = "Server=.;Database=MyDb;Trusted_Connection=True;";
    }

    public void Print() => Console.WriteLine($"Connection: {ConnectionString}");
}

class Program
{
    static void Main()
    {
        var a = AppConfig.Instance;
        var b = AppConfig.Instance;
        Console.WriteLine(Object.ReferenceEquals(a, b)); // True
        a.Print();
    }
}

File: 1. Creational Design Patterns/Singleton/Typescript/SingletonExample.ts
// Singleton Pattern Example in TypeScript
// Node-friendly, no external typings required.

declare const console: any;

declare const global: any;

class AppConfig {
  private static _instance: AppConfig | null = null;
  public readonly connectionString: string;

  private constructor() {
    this.connectionString = 'redis://localhost:6379';
  }

  public static getInstance(): AppConfig {
    if (!AppConfig._instance) {
      AppConfig._instance = new AppConfig();
    }
    return AppConfig._instance;
  }

  public print(): void {
    console.log('Connection:', this.connectionString);
  }
}

const a = AppConfig.getInstance();
const b = AppConfig.getInstance();
console.log(a === b); // true
a.print();
